lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs

```diff
@@ -1738,41 +1739,48 @@ selectCoins
+    era <- liftIO $ NW.currentNodeEra netLayer
+    recentEra@(AnyRecentEra (_ :: WriteTx.RecentEra e)) <- guardIsRecentEra era
+    withWorkerCtx ctx walletId liftE liftE $ \workerCtx -> do
+        let db = workerCtx ^. dbLayer
+            ti = timeInterpreter netLayer
+        pp <- liftIO $ NW.currentProtocolParameters (workerCtx ^. networkLayer)
+        withdrawal <-
+            body ^. #withdrawal
+                & maybe (pure NoWithdrawal)
+                    (shelleyOnlyMkWithdrawal @s @k @n
+                        netLayer txLayer db walletId era)
+
+        let paymentOuts = NE.toList $ addressAmountToTxOut <$> body ^. #payments
         let txCtx = defaultTransactionCtx
-                { txWithdrawal = wdrl
+                { txWithdrawal = withdrawal
                 , txMetadata = getApiT <$> body ^. #metadata
                 }
-        let transform s sel =
-                W.assignChangeAddresses genChange sel s
-                & uncurry (W.selectionToUnsignedTx (txWithdrawal txCtx))
-        (utxoAvailable, wallet, pendingTxs) <-
-            liftHandler $ W.readWalletUTxOIndex @_ @s @k wrk wid
-        pp <- liftIO $ NW.currentProtocolParameters (wrk ^. networkLayer)
-        let selectAssetsParams = W.SelectAssetsParams
-                { outputs = F.toList outs
-                , pendingTxs
-                , randomSeed = Nothing
-                , txContext = txCtx
-                , utxoAvailableForInputs = UTxOSelection.fromIndex utxoAvailable
-                , utxoAvailableForCollateral = UTxOIndex.toMap utxoAvailable
-                , wallet
-                , selectionStrategy = SelectionStrategyOptimal
-                }
-        utx <- liftHandler $
-            W.selectAssets @_ @_ @s @k @'CredFromKeyK
-                wrk era pp selectAssetsParams transform
-        pure $ mkApiCoinSelection [] [] Nothing md utx
+
+        (cardanoTx, walletState) <- liftIO $ W.buildTransaction @s @k @n @e
+            recentEra db txLayer ti walletId argGenChange pp txCtx paymentOuts
+
+        let W.CoinSelection{..} =
+                W.buildCoinSelectionForTransaction @s @k @n
+                    walletState
+                    paymentOuts
+                    (W.stakeKeyDeposit pp)
+                    Nothing -- delegation action
+                    cardanoTx
+
+        pure ApiCoinSelection
+            { inputs = mkApiCoinSelectionInput <$> inputs
+            , outputs = mkApiCoinSelectionOutput <$> outputs
+            , change = mkApiCoinSelectionChange <$> change
+            , collateral = mkApiCoinSelectionCollateral <$> collateral
+            , certificates = uncurry mkApiCoinSelectionCerts <$>
+                delegationAction
+            , withdrawals = mkApiCoinSelectionWithdrawal <$> withdrawals
+            , depositsTaken = maybeToList $ mkApiCoin <$> deposit
+            , depositsReturned = maybeToList $ mkApiCoin <$> refund
+            , metadata = Nothing
```

Comment:
☝️ We need to still return the metadata

It used to be:
```haskell
pure $ 
    mkApiCoinSelection 
        [] 
        [] 
        Nothing -- delegation action
        md -- metadata
        utx

```

Reply:
Good catch!